src/object.d

```diff
@@ -424,6 +424,547 @@ class TypeInfo
+                arg1 = cast() __typeid!(void*); // TODO: revisit
+                return 0;
+            }
+        else static if (is(T == __vector(U2), U2))
+            override int argTypes(out TypeInfo arg1, out TypeInfo arg2) @nogc @trusted
+            {
+                return (cast() __typeid!U2).argTypes(arg1, arg2);
+            }
+    }
+
+    // The postblit API (static and dynamic)
+    static if (is(typeof(&T.__postblit)))
+        static void postblit(UnqualifiedType* p)
+        {
+            p.__postblit();
+        }
+    else static if (__traits(isStaticArray, T) && is(typeof(&UnqualifiedType.init[0].__postblit)))
+        static void postblit(UnqualifiedType* p)
+        {
+            foreach (i, ref e; *p)
+                e.__postblit();
+        }
+    else
+        alias postblit = doNothing;
+
+    // The destroy API (static and dynamic)
+    static if (is(typeof(&T.__dtor)))
+        static void destroy(UnqualifiedType* p)
+        {
+            p.__dtor();
+        }
+    else static if (__traits(isStaticArray, T) && is(typeof(&UnqualifiedType.init[0].__dtor)))
+        static void destroy(UnqualifiedType* p)
+        {
+            foreach_reverse (i, ref e; *p)
+                e.__dtor();
+        }
+    else
+        alias destroy = doNothing;
+
+    nothrow: pure:
+
+    private final void doNothing(UnqualifiedType*) @nogc @safe {}
+
+    // On the off chance someone calls typeid(T) == typeid(U)
+    final bool opEquals(U)(const TypeInfoImpl!U rhs) @nogc @safe
+    {
+        pragma(inline, true);
+        return is(T == U);
+    }
+
+    // Is this equal to another TypeInfo?
+    // WARNING: relies on a 1:1 mapping Type : TypeInfoImpl!Type
+    override bool opEquals(const Object rhs) @nogc @safe
+    {
+        return this is rhs
+            // Instance may be duplicated across DLLs, but has the same type.
+            || cast(const TypeInfoImpl) rhs !is null;
+    }
+
+    // On the off chance someone calls typeid(T) < typeid(U) etc.
+    final int opCmp(U)(const TypeInfoImpl!U rhs) @nogc @safe
+    {
+        pragma(inline, true);
+        return is(T == U) ? 0 : __cmp(T.stringof, U.stringof);
+    }
+
+    // Is this less than/equal/greater than another TypeInfo?
+    // WARNING: relies on a 1:1 mapping Type : TypeInfoImpl!Type
+    override int opCmp(const Object rhs) @safe
+    {
+        if (this is rhs
+                // Instance may be duplicated across DLLs, but has the same type.
+                || cast(const TypeInfoImpl) rhs !is null)
+            return 0;
+        if (auto ti = cast(const TypeInfo) rhs)
+            return __cmp(this.toString, ti.toString);
+        // TypeInfo objects are all "greater" than all other types.
+        // TODO: devise a principled way to define/justify this.
+        return 1;
+    }
+
+    override string toString() @nogc @safe { return T.stringof; }
+
+    final static size_t getHash(scope const T* p) @nogc
+    {
+        pragma(inline, true);
```

Comment:
How about implementing this function as
```d
static if (__traits(isIntegral, T) && is(T : size_t))
    return *p;
else
    return hashOf(*p);
```

Reply:
Was trying to do "the right thing", which is that a slight change in the integral should affect all bits in the output. So I went with the Knuth hash, which is a compromise between speed and quality. If compatibility is paramount, we can always simplify.

Reply:
My thinking (which may be wrong-headed) is that the primary purpose of `TypeInfo.getHash` is catering to built-in associative arrays. The `calcHash` function used by built-in AAs is already [using MurmurHash2's bit avalanche function](https://github.com/dlang/druntime/blob/104ac712331e4d3573fc277084334a528b5dadb1/src/rt/aaA.d#L446-L455) to transform the output of `TypeInfo.getHash`.

Reply:
Cool. At any rate, here's hoping that AAs will be templated and rid of TypeInfo.

Reply:
If we should do something different with integers, shouldn't `hashOf` be changed instead, and have this forward to it?

The point of `hashOf` (which is used by `TypeInfo`, but does not require `TypeInfo`) is to have a consistent mechanism to hash all types. I think we should use the existing mechanism.

Alternately, we can eliminate `hashOf`, and have it forward to `__typeid!T.getHash`.

I don't disagree with the change, as I have no specific thoughts on hashing, I always use what others have provided. But there shouldn't be multiple ways that druntime hashes things.

Reply:
@schveiguy yes, mos def. A unified hashing scheme that all code uses is best. Will look into it when I get out of the woods wrt functionality.

Reply:
@n8sh @schveiguy I reworked the code to use `getHash`. Any work on that can be done later.